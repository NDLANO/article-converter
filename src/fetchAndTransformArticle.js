import cheerio from 'cheerio';
import { fetchArticle } from './api/articleApi';
import { transform } from './transformers';
import config from './config';
import { getCopyString } from './plugins/pluginHelpers';

export async function transformArticle(
  article,
  lang,
  accessToken,
  options = {}
) {
  const articleContent = article.content.content
    ? cheerio.load(article.content.content, {
        recognizeSelfClosing: true,
      })
    : undefined;
  const { html, embedMetaData } = articleContent
    ? await transform(
        articleContent,
        lang,
        accessToken,
        article.visualElement,
        options
      )
    : {};
  const copyText = getCopyString(
    article.title.title,
    config.ndlaFrontendDomain,
    options.path,
    article.copyright,
    lang
  );

  return {
    ...article,
    content: html || '',
    metaData: { ...embedMetaData, copyText } || '',
    title: article.title.title || '',
    tags: article.tags ? article.tags.tags : [],
    introduction: article.introduction
      ? article.introduction.introduction
      : undefined,
    metaDescription: article.metaDescription
      ? article.metaDescription.metaDescription
      : '',
  };
}

export default async function fetchAndTransformArticle(
  articleId,
  lang,
  accessToken,
  options = {}
) {
  const article = await fetchArticle(articleId, accessToken, lang);
  const transformedArticle = await transformArticle(
    article,
    lang,
    accessToken,
    options
  );
  return transformedArticle;
}
