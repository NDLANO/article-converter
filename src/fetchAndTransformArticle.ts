import cheerio from 'cheerio';
import { IArticleV2 } from '@ndla/types-article-api';
import { fetchArticle } from './api/articleApi';
import { transform } from './transformers';
import config from './config';
import { getCopyString } from './plugins/pluginHelpers';
import { LocaleType, ResponseHeaders, TransformedArticle, TransformOptions } from './interfaces';

export async function transformArticle(
  article: IArticleV2,
  headers: ResponseHeaders,
  lang: LocaleType,
  accessToken: string,
  feideToken: string,
  options: TransformOptions = {},
): Promise<TransformedArticle> {
  const articleContent = article.content.content
    ? cheerio.load(article.content.content, {
        recognizeSelfClosing: true,
      })
    : undefined;

  const { html, embedMetaData, responseHeaders } = articleContent
    ? await transform(
        articleContent,
        headers,
        lang,
        accessToken,
        feideToken,
        article.visualElement,
        options,
      )
    : { html: undefined, embedMetaData: undefined, responseHeaders: undefined };

  const htmlString: string = html ?? '';

  const copyText: string = getCopyString(
    article.title.title,
    config.ndlaFrontendDomain,
    options.path,
    article.copyright,
    lang,
  );
  const hasContent = article.articleType === 'standard' || cheerio.load(htmlString).text() !== '';

  return {
    ...article,
    content: hasContent ? htmlString : '',
    metaData: { ...embedMetaData, copyText } || '',
    title: article.title.title || '',
    tags: article.tags ? article.tags.tags : [],
    introduction: article.introduction ? article.introduction.introduction : undefined,
    metaDescription: article.metaDescription ? article.metaDescription.metaDescription : '',
    headerData: responseHeaders ?? {},
  };
}

export default async function fetchAndTransformArticle(
  articleId: string,
  lang: LocaleType,
  accessToken: string,
  feideToken: string,
  options = {},
): Promise<TransformedArticle> {
  const [article, headers] = await fetchArticle(articleId, accessToken, feideToken, lang);
  return await transformArticle(article, headers, lang, accessToken, feideToken, options);
}
