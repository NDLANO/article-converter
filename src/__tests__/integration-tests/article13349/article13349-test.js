/**
 * Copyright (c) 2016-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import nock from 'nock';
import { prettify } from '../../testHelpers';
import article13349 from './article13349';

import fetchAndTransformArticle from '../../../fetchAndTransformArticle';

// Tests that we can have a filelist inside aside
test('app/fetchAndTransformArticle 13349', async () => {
  nock('http://ndla-api')
    .get('/article-api/v2/articles/13349?language=nb&fallback=true')
    .reply(200, article13349);
  nock('http://ndla-api').head('/files/82255/Gant%20diagram.pdf').reply(200);

  const transformed = await fetchAndTransformArticle('13349', {
    lang: 'nb',
    accessToken: 'some_token',
    feideToken: 'some_other_token',
  });
  const { content, ...rest } = transformed;

  expect(rest).toMatchSnapshot();
  expect(prettify(content)).toMatchSnapshot();
});
