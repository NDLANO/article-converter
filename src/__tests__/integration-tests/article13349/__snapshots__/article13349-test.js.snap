// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app/fetchAndTransformArticle 13349 1`] = `
Object {
  "articleType": "standard",
  "copyright": Object {
    "creators": Array [
      Object {
        "name": "Steinar Olsen",
        "type": "Writer",
      },
    ],
    "license": Object {
      "description": "Creative Commons Attribution-ShareAlike 4.0 International",
      "license": "CC-BY-SA-4.0",
      "url": "https://creativecommons.org/licenses/by-sa/4.0/",
    },
    "origin": "",
    "processors": Array [],
    "rightsholders": Array [],
  },
  "created": "2009-09-15T07:20:32Z",
  "headerData": Object {},
  "id": 13349,
  "introduction": "Her skal du installere og lære om forskjellige lyskilder styrt av de mest vanlige lysbryterne som blir brukt i boliger. Du skal lære deg åpen og skjult installasjon. Samtidig skal du lære om standarder og forskrifter som finnes i boliginstallasjon",
  "metaData": Object {
    "copyText": "Forfatter: Steinar Olsen. Planlegging [Internett]. Hentet fra: https://test.ndla.no Lest: 19.05.2021",
  },
  "metaDescription": "Her skal du installere og lære om forskjellige lyskilder styrt av de mest vanlige lysbryterne som blir brukt i boliger.",
  "metaImage": Object {
    "alt": "",
    "language": "nb",
    "url": "https://staging.api.ndla.no/image-api/raw/id/3948",
  },
  "oldNdlaUrl": "//red.ndla.no/node/4901",
  "requiredLibraries": Array [],
  "revision": 1,
  "supportedLanguages": Array [
    "nb",
    "nn",
  ],
  "tags": Array [
    "forberedelse",
    "lysstyring",
    "planlegging",
  ],
  "title": "Planlegging",
  "updated": "2018-02-20T14:58:39Z",
  "updatedBy": "r0gHb9Xg3li4yyXv0QSGQczV3bviakrT",
}
`;

exports[`app/fetchAndTransformArticle 13349 2`] = `
"<section>
  <aside class=\\"c-aside c-aside--wideScreen\\">
    <div class=\\"c-aside__content\\">
      <section class=\\"c-file-list\\">
        <h1 class=\\"c-file-list__heading\\">Filer</h1>
        <ul class=\\"c-file-list__files\\">
          <li class=\\"c-file-list__item\\">
            <a
              href=\\"https://staging.api.ndla.no/files/82255/Gant diagram.pdf\\"
              class=\\"c-file-list__link\\"
              target=\\"_blank\\"
              aria-label=\\"Gantt-diagram (PDF)\\"
              aria-describedby=\\"testid_pdf\\"
              ><svg
                fill=\\"currentColor\\"
                preserveAspectRatio=\\"xMidYMid meet\\"
                height=\\"1em\\"
                width=\\"1em\\"
                class=\\"c-icon\\"
                viewBox=\\"0 0 24 24\\"
                data-license=\\"Apache License 2.0\\"
                data-source=\\"Material Design\\"
                style=\\"vertical-align: middle\\"
              >
                <title>Download</title>
                <g>
                  <path d=\\"M19 9h-4V3H9v6H5l7 7 7-7zM5 18v2h14v-2H5z\\"></path>
                  <path d=\\"M0 0h24v24H0z\\" fill=\\"none\\"></path>
                </g></svg
              ><span class=\\"c-file-list__link-text\\"
                ><span>Gantt-diagram (PDF)</span></span
              ><span
                class=\\"c-file-list__tooltip\\"
                aria-hidden=\\"true\\"
                role=\\"tooltip\\"
                id=\\"testid_pdf\\"
                ><span class=\\"c-file-list__tooltip-text\\"
                  >Last ned fil: Gant diagram.pdf</span
                ></span
              ></a
            >
          </li>
        </ul>
      </section>
      <p><strong>Lenke:</strong></p>
      <p>
        <a
          href=\\"http://no.wikipedia.org/wiki/Norsk_Standard\\"
          rel=\\"noopener noreferrer\\"
          target=\\"_blank\\"
          title=\\"Norsk Standard\\"
          >Norsk Standard</a
        >
      </p>
    </div>
  </aside>
  <style data-emotion-css=\\"1ly2gtx-ContainerDiv\\">
    .css-1ly2gtx-ContainerDiv:hover button {
      cursor: pointer;
      opacity: 0.5;
    }
    .css-1ly2gtx-ContainerDiv h2 {
      position: relative;
      left: -2em;
    }
  </style>
  <div
    data-header-copy-container=\\"true\\"
    data-title=\\"I planleggingsdelen må du:\\"
    class=\\"css-1ly2gtx-ContainerDiv e18czas71\\"
  >
    <style data-emotion-css=\\"1palhdc-IconButton\\">
      .css-1palhdc-IconButton {
        float: left;
        position: relative;
        left: -3em;
        top: 0.1em;
        background: none;
        border: 0;
        z-index: 1;
        -webkit-transition: 0.2s;
        transition: 0.2s;
        opacity: 0;
      }
      .css-1palhdc-IconButton svg {
        width: 30px;
        height: 30px;
      }</style
    ><button
      data-title=\\"I-planleggingsdelen-m%C3%A5-du%3A\\"
      class=\\"css-1palhdc-IconButton e18czas70\\"
    >
      <style data-emotion-css=\\"1twu449-TooltipWrapper\\">
        .css-1twu449-TooltipWrapper {
          position: relative;
        }
      </style>
      <div class=\\"css-1twu449-TooltipWrapper exsy4zb0\\">
        <style data-emotion-css=\\"1gms6h4-Fade\\">
          .css-1gms6h4-Fade {
            opacity: 0;
            position: absolute;
            z-index: 9999;
            pointer-events: none;
            -webkit-animation-fill-mode: forwards;
            animation-fill-mode: forwards;
            -webkit-animation-delay: 0ms;
            animation-delay: 0ms;
            -webkit-animation-duration: 300ms;
            animation-duration: 300ms;
          }
          @-webkit-keyframes fadeInTooltip {
            0% {
              opacity: 0;
            }
            100% {
              opacity: 1;
            }
          }
          @keyframes fadeInTooltip {
            0% {
              opacity: 0;
            }
            100% {
              opacity: 1;
            }
          }
          @-webkit-keyframes fadeOutTooltip {
            0% {
              opacity: 1;
            }
            100% {
              opacity: 0;
            }
          }
          @keyframes fadeOutTooltip {
            0% {
              opacity: 1;
            }
            100% {
              opacity: 0;
            }
          }
        </style>
        <div class=\\"css-1gms6h4-Fade exsy4zb1\\">
          <style data-emotion-css=\\"aztq0g-tooltipCss\\">
            .css-aztq0g-tooltipCss{display:block;color:#fff;background:#444;border-radius:2px;padding:6.5px 13px;font-family:'Source Sans Pro',Helvetica,Arial,STKaiti,'华文楷体',KaiTi,SimKai,'楷体',KaiU,DFKai-SB,'標楷體',SongTi,'宋体',sans-serif;font-size:14;font-size:0.7777777777777778rem;line-height:1.2 font-weight:400;text-align:center;white-space:nowrap;max-width:calc(100vw - #{26px});}</style
          ><span role=\\"tooltip\\" class=\\"css-aztq0g-tooltipCss\\"
            >Kopier lenke til overskriften</span
          >
        </div>
        <div aria-label=\\"Kopier lenke til overskriften\\" class=\\"\\">
          <svg
            fill=\\"currentColor\\"
            preserveAspectRatio=\\"xMidYMid meet\\"
            height=\\"1em\\"
            width=\\"1em\\"
            class=\\"c-icon\\"
            viewBox=\\"0 0 24 24\\"
            data-license=\\"Apache License 2.0\\"
            data-source=\\"Material Design\\"
            style=\\"vertical-align: middle\\"
          >
            <g>
              <path d=\\"M0 0h24v24H0z\\" fill=\\"none\\"></path>
              <path
                d=\\"M3.9 12c0-1.71 1.39-3.1 3.1-3.1h4V7H7c-2.76 0-5 2.24-5 5s2.24 5 5 5h4v-1.9H7c-1.71 0-3.1-1.39-3.1-3.1zM8 13h8v-2H8v2zm9-6h-4v1.9h4c1.71 0 3.1 1.39 3.1 3.1s-1.39 3.1-3.1 3.1h-4V17h4c2.76 0 5-2.24 5-5s-2.24-5-5-5z\\"
              ></path>
            </g>
          </svg>
        </div>
      </div>
    </button>
    <h2 id=\\"I-planleggingsdelen-m%C3%A5-du%3A\\" tabindex=\\"0\\">
      I planleggingsdelen må du:
    </h2>
  </div>
  <p>
    <strong>1.</strong> Foreta en risikovurdering, der du vurderer risikoen ved
    jobben du skal gjøre (HMS, vernetiltak, forskrifter), samt risikoen når
    installasjonen er i bruk.
  </p>
  <ul>
    <li>Du skal arbeide i høyden. Hva må vi tenke på her?</li>
    <li>Du skal kanskje jobbe med spenning?</li>
    <li>Tar vi hensyn til sikkerhetsforskriften?</li>
    <li>Er det verktøy du ikke kjenner, og som du må ha opplæring i?</li>
  </ul>
  <p>
    <strong>2.</strong> Utarbeide en materialliste, slik at du får med deg
    utstyr og materiell du trenger for å utføre jobben.
  </p>
  <ul>
    <li>Er installasjonen åpen eller skjult?</li>
    <li>Finne fram egnet liste, og finne leverandør- eller grossistkatalog.</li>
  </ul>
  <p>
    <strong>3.</strong> Lage en plan for arbeidet (Gant-diagram), det vil si at
    du utarbeider en framdriftsplan der du setter opp på papir i kronologisk
    rekkefølge det du skal gjøre.
  </p>
  <ul>
    <li>
      Gå gjennom oppdraget, og skriv ned de jobbene du skal gjennom i perioden
      (i venstre kolonne).
    </li>
    <li>
      Sett av den tiden du mener du trenger til de forskjellige delene av
      arbeidet.
    </li>
    <li>
      Lag deg en \\"buffer\\" på slutten av oppdraget slik at du har tid til
      ferdigstilling, selv om ikke alt går helt etter planen.
    </li>
  </ul>
  <p>
    Finne fram aktuelt verktøy for jobben du skal gjøre. Klargjøre aktuell
    anleggsdokumentasjon for jobben. I dette oppdraget er det
    installasjonstegning og koblingsskjema.
  </p>
  <aside class=\\"c-aside c-aside--narrowScreen\\">
    <div class=\\"c-aside__content\\">
      <section class=\\"c-file-list\\">
        <h1 class=\\"c-file-list__heading\\">Filer</h1>
        <ul class=\\"c-file-list__files\\">
          <li class=\\"c-file-list__item\\">
            <a
              href=\\"https://staging.api.ndla.no/files/82255/Gant diagram.pdf\\"
              class=\\"c-file-list__link\\"
              target=\\"_blank\\"
              aria-label=\\"Gantt-diagram (PDF)\\"
              aria-describedby=\\"testid_pdf\\"
              ><svg
                fill=\\"currentColor\\"
                preserveAspectRatio=\\"xMidYMid meet\\"
                height=\\"1em\\"
                width=\\"1em\\"
                class=\\"c-icon\\"
                viewBox=\\"0 0 24 24\\"
                data-license=\\"Apache License 2.0\\"
                data-source=\\"Material Design\\"
                style=\\"vertical-align: middle\\"
              >
                <title>Download</title>
                <g>
                  <path d=\\"M19 9h-4V3H9v6H5l7 7 7-7zM5 18v2h14v-2H5z\\"></path>
                  <path d=\\"M0 0h24v24H0z\\" fill=\\"none\\"></path>
                </g></svg
              ><span class=\\"c-file-list__link-text\\"
                ><span>Gantt-diagram (PDF)</span></span
              ><span
                class=\\"c-file-list__tooltip\\"
                aria-hidden=\\"true\\"
                role=\\"tooltip\\"
                id=\\"testid_pdf\\"
                ><span class=\\"c-file-list__tooltip-text\\"
                  >Last ned fil: Gant diagram.pdf</span
                ></span
              ></a
            >
          </li>
        </ul>
      </section>
      <p><strong>Lenke:</strong></p>
      <p>
        <a
          href=\\"http://no.wikipedia.org/wiki/Norsk_Standard\\"
          rel=\\"noopener noreferrer\\"
          target=\\"_blank\\"
          title=\\"Norsk Standard\\"
          >Norsk Standard</a
        >
      </p>
    </div>
  </aside>
</section>
"
`;
